dev/TeachingTip/InteractionTests/TeachingTipTests.cs

```diff
@@ -0,0 +1,820 @@
+                elements.targetWidthTextBlock = new TextBlock(FindElement.ByName("TargetWidthTextBlock"));
+                Verify.IsNotNull(elements.targetWidthTextBlock);
+            }
+
+            if (elements.targetHeightTextBlock == null)
+            {
+                Log.Comment("Find the TargetHeightTextBlock");
+                elements.targetHeightTextBlock = new TextBlock(FindElement.ByName("TargetHeightTextBlock"));
+                Verify.IsNotNull(elements.targetHeightTextBlock);
+            }
+
+            elements.getTargetBoundsButton.Invoke();
+
+            var retVal = new Vector4();
+            retVal.W = (int)Math.Floor(double.Parse(elements.targetXOffsetTextBlock.GetText()));
+            retVal.X = (int)Math.Floor(double.Parse(elements.targetYOffsetTextBlock.GetText()));
+            retVal.Y = (int)Math.Floor(double.Parse(elements.targetWidthTextBlock.GetText()));
+            retVal.Z = (int)Math.Floor(double.Parse(elements.targetHeightTextBlock.GetText()));
+            return retVal;
+        }
+
+        private string GetEffectivePlacement()
+        {
+            if (elements.effectivePlacementTextBlock == null)
+            {
+                Log.Comment("Find the EffectivePlacementTextBlock");
+                elements.effectivePlacementTextBlock = new TextBlock(FindElement.ByName("EffectivePlacementTextBlock"));
+                Verify.IsNotNull(elements.effectivePlacementTextBlock);
+            }
+            return elements.effectivePlacementTextBlock.GetText();
+        }
+
+        // The test UI has a list box which the teaching tip populates with messages about which events have fired and other useful
+        // Debugging info. This method returns the message at the provided index, which helps testing that events were received in
+        // the expected order.
+        private ListBoxItem GetTeachingTipDebugMessage(int index)
+        {
+            if (elements.lstTeachingTipEvents == null)
+            {
+                Log.Comment("Find the lstTeachingTipEvents");
+                elements.lstTeachingTipEvents = new ListBox(FindElement.ByName("lstTeachingTipEvents"));
+                Verify.IsNotNull(elements.lstTeachingTipEvents);
+            }
+            return elements.lstTeachingTipEvents.Items[index];
+        }
+
+        private bool WaitForChecked(CheckBox checkBox, double millisecondsTimeout = 2000, bool throwOnError = true)
+        {
+            return WaitForCheckBoxUpdated(checkBox, ToggleState.On, millisecondsTimeout, throwOnError);
+        }
+
+        private bool WaitForUnchecked(CheckBox checkBox, double millisecondsTimeout = 2000, bool throwOnError = true)
+        {
+            return WaitForCheckBoxUpdated(checkBox, ToggleState.Off, millisecondsTimeout, throwOnError);
+        }
+
+        private bool WaitForCheckBoxUpdated(CheckBox checkBox, ToggleState state, double millisecondsTimeout, bool throwOnError)
```

Comment:
This seems generally useful. Consider moving it into a general test helper class such as WaitHelper.

Reply:
I wrote something similar for the snap points tests, I think it would be a good idea to try and consolidate both of them.

---
In reply to: [252903422](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r252903422) [](ancestors = 252903422)

Reply:
I'd like to do this in a follow up CR though, this one is pretty unwieldy as it is

---
In reply to: [252914198](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r252914198) [](ancestors = 252914198,252903422)